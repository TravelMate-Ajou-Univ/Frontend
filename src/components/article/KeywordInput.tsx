"use client";

import { useState } from "react";
import PlusIcon from "../ui/icons/PlusIcon";
import DropDownList from "../ui/DropDown/DropDownList";
import useOutSideRef from "@/hooks/useClickOutside";
import { getKeywords } from "@/service/axios/article";
import { KeywordType } from "@/model/article";

interface Props {
  addKeyword: (keyword: string) => void;
}

export default function KeywordInput({ addKeyword }: Props) {
  const [keyword, setKeyword] = useState<string>("");
  const [searchedKeyword, setSearchedKeyword] = useState<string[]>([]);
  const [dropdown, setDropdown] = useState<boolean>(false);
  const ref = useOutSideRef(() => setDropdown(false));

  const addKeywordList = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    addKeyword(keyword);
    setKeyword("");
  };

  const handleKeyword = async (keyword: string) => {
    setKeyword(keyword);
    if (keyword === "") {
      setDropdown(false);
      return;
    }
    const keywords = await getKeywords(keyword);
    if (keywords) {
      setSearchedKeyword(keywords.slice(0, 5).map((item: any) => item.name));
      setDropdown(true);
    }
  };

  const handleDropdownKeyword = (keyword: string) => {
    setKeyword(keyword);
    setDropdown(false);
  };

  return (
    <div className="flex flex-col">
      <form className="flex" onSubmit={e => addKeywordList(e)}>
        <input
          className="focus:outline-none border-b px-2 py-1 text-sm w-80"
          type="text"
          placeholder="#맛집, #카페, #숙소 등의 키워드를 입력해주세요 :)"
          value={keyword}
          onChange={e => handleKeyword(e.target.value)}
        />
        <button className="border-b text-sm text-primary w-8 h-8" type="submit">
          <PlusIcon noBorder />
        </button>
      </form>
      {dropdown && searchedKeyword.length !== 0 && (
        <div className="w-80 relative" ref={ref}>
          <DropDownList
            list={searchedKeyword}
            setSelected={handleDropdownKeyword}
            isSearchWord
            isKeyword
          />
        </div>
      )}
    </div>
  );
}
