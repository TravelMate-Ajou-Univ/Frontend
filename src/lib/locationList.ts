export const locationList = [
  "서울",
  "경기/인천",
  "강원",
  "충청/대전",
  "전라/광주",
  "경북/대구",
  "경남/울산/부산",
  "제주"
];
