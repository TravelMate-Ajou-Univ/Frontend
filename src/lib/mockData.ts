export const articleMock = {
  id: 1,
  title: "",
  thumbnail: null,
  location: "경기/인천",
  authorId: 1,
  springVersionID: 1,
  summerVersionID: null,
  fallVersionID: null,
  winterVersionID: null,
  articleTagMap: [
    {
      id: 228,
      articleId: 20,
      tagId: 8,
      tag: {
        id: 8,
        name: "카페",
        createdAt: "2023-11-05T11:13:46.000Z",
        updatedAt: "2023-11-05T11:13:46.000Z"
      }
    },
    {
      id: 229,
      articleId: 20,
      tagId: 104,
      tag: {
        id: 104,
        name: "행궁",
        createdAt: "2023-12-08T07:49:33.000Z",
        updatedAt: "2023-12-08T07:49:33.000Z"
      }
    },
    {
      id: 230,
      articleId: 20,
      tagId: 60,
      tag: {
        id: 60,
        name: "행궁 카페",
        createdAt: "2023-11-23T05:46:55.000Z",
        updatedAt: "2023-11-23T05:46:55.000Z"
      }
    }
  ],
  spring: null,
  summer: {
    id: 1,
    articleId: 1,
    userId: 1,
    period: "SUMMER",
    content: "<p>hi</p>",
    createdAt: "2023-12-02T09:31:38.000Z",
    updatedAt: "2023-12-02T09:31:38.000Z"
  },
  fall: null,
  winter: null,
  articleBookmarkMap: [
    {
      id: 90,
      period: "WINTER",
      articleId: 20,
      bookmarkId: 227,
      createdAt: "2023-12-20T14:07:44.000Z",
      updatedAt: "2023-12-20T14:07:44.000Z",
      bookmark: {
        id: 227,
        content: "4층까지 있고 넓고 쾌적",
        locationId: 90,
        userId: 3,
        createdAt: "2023-12-20T14:02:38.000Z",
        deletedAt: null,
        location: {
          id: 90,
          latitude: "37.2848574",
          longitude: "127.0144045",
          placeId: "ChIJ9_ndU51DezURyWyJWMNsnKs"
        }
      }
    }
  ],
  isFavorite: false
};

export const searchedKeywordMock = [
  {
    id: 2,
    name: "제주 맛집",
    createdAt: "2023-11-05T11:00:18.000Z",
    updatedAt: "2023-11-05T11:00:18.000Z"
  },
  {
    id: 34,
    name: "맛집",
    createdAt: "2023-11-08T12:43:01.000Z",
    updatedAt: "2023-11-08T12:43:01.000Z"
  }
];
