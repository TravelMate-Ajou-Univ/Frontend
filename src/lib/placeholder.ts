const KEYWORD_INPUT = "#맛집, #카페, #숙소 등의 키워드를 입력해주세요 :)";

export const placeholder = {
  keywordInput: KEYWORD_INPUT
};
