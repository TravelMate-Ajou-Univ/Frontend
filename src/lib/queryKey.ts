const ARTICLE = "article";
const ARTICLE_EDIT = "article-edit";
const ARTICLE_COUNT = "article-count";

const KEYWORD = "keyword";

export const queryKey = {
  article: ARTICLE,
  articleEdit: ARTICLE_EDIT,
  articleCount: ARTICLE_COUNT,
  keyword: KEYWORD
};
