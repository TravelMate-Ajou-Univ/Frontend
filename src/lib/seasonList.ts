export const seasonList = ["봄", "여름", "가을", "겨울"];

export const seasonMapper = {
  봄: "SPRING",
  여름: "SUMMER",
  가을: "FALL",
  겨울: "WINTER"
};
