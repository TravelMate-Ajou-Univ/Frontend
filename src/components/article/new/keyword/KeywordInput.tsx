"use client";

import { useEffect, useState } from "react";
import PlusIcon from "@/components/ui/icons/PlusIcon";
import DropDownList from "@/components/ui/dropDown/DropDownList";
import useOutSideRef from "@/hooks/useClickOutside";
import { useGetKeywordsQuery } from "@/service/react-query/keyword";
import { KeywordType } from "@/model/article";
import { placeholder } from "@/lib/placeholder";

export interface Props {
  inputId: string;
  addKeyword: (keyword: string) => void;
  disabled?: boolean;
}

export default function KeywordInput({
  inputId,
  addKeyword,
  disabled = false
}: Props) {
  const [keyword, setKeyword] = useState<string>("");
  const [dropdown, setDropdown] = useState<boolean>(false);
  const ref = useOutSideRef(() => setDropdown(false));

  const { data: keywords, refetch } = useGetKeywordsQuery(keyword, setDropdown);

  useEffect(() => {
    refetch();
  }, [keyword, refetch]);

  const addKeywordList = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    addKeyword(keyword);
    setKeyword("");
  };

  const handleKeyword = (e: React.ChangeEvent<HTMLInputElement>) => {
    const keyword = e.target.value;
    setKeyword(keyword);
    if (keyword === "") setDropdown(false);
  };

  const handleDropdownKeyword = (keyword: string) => {
    addKeyword(keyword);
    setKeyword("");
    setDropdown(false);
  };

  return (
    <div className="flex flex-col">
      <form className="flex" onSubmit={addKeywordList}>
        <input
          id={inputId}
          className="focus:outline-none border-b px-2 py-1 text-sm md:w-80 sm:w-72 w-64 rounded-none"
          type="text"
          placeholder={placeholder.keywordInput}
          value={keyword}
          onChange={handleKeyword}
          disabled={disabled}
        />
        <button className="border-b text-sm text-primary w-8 h-8" type="submit">
          <PlusIcon noBorder />
        </button>
      </form>
      {dropdown && keywords && keywords.length !== 0 && (
        <div className="md:w-80 sm:w-72 w-64 relative" ref={ref}>
          <DropDownList
            list={keywords.slice(0, 5).map((item: KeywordType) => item.name)}
            setSelected={handleDropdownKeyword}
            isSearchWord
            isKeyword
          />
        </div>
      )}
    </div>
  );
}
