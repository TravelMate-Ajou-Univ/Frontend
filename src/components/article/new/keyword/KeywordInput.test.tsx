import { composeStories } from "@storybook/react";
import * as stories from "./KeywordInput.stories";
import { axiosMock } from "@/test/axios-mock";
import { searchedKeywordMock } from "@/lib/mockData";
import customRender from "@/test/customRender";
import { fireEvent, waitFor } from "@testing-library/react";

const KEYWORD = "맛집";

const { Default: KeywordInput } = composeStories(stories);

beforeEach(() => {
  axiosMock.onGet(`tags/search/${KEYWORD}`).reply(200, searchedKeywordMock);
});

afterEach(() => {
  axiosMock.reset();
});

describe("<KeywordInput />", () => {
  it("키워드 입력 시 검색된 키워드를 보여준다.", async () => {
    const component = customRender(<KeywordInput />);

    const input = component.getByRole("textbox");
    fireEvent.change(input, { target: { value: KEYWORD } });

    await waitFor(() => {
      expect(component.getByRole("list")).toBeInTheDocument();

      expect(
        component.getByRole("listitem", {
          name(_, element) {
            return element.textContent === "제주 맛집";
          }
        })
      ).toBeInTheDocument();
      expect(
        component.getByRole("listitem", {
          name(_, element) {
            return element.textContent === "맛집";
          }
        })
      ).toBeInTheDocument();
    });

    expect(component).toMatchSnapshot();
  });
});
