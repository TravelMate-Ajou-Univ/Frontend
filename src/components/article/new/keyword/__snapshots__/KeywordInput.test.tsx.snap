// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeywordInput /> 키워드 입력 시 검색된 키워드를 보여준다. 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col"
      >
        <form
          class="flex"
        >
          <input
            class="focus:outline-none border-b px-2 py-1 text-sm md:w-80 sm:w-72 w-64 rounded-none"
            id="keywordInput"
            placeholder="#맛집, #카페, #숙소 등의 키워드를 입력해주세요 :)"
            type="text"
            value="맛집"
          />
          <button
            class="border-b text-sm text-primary w-8 h-8"
            type="submit"
          >
            <svg
              class="w-full h-full p-2 false"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8Z"
              />
              <path
                d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8Z"
              />
            </svg>
          </button>
        </form>
        <div
          class="md:w-80 sm:w-72 w-64 relative"
        >
          <ul
            class="absolute left-0 w-full border rounded-md bg-white z-50 
      top-0"
          >
            <li
              class="py-1 hover:bg-gray-200 px-3 text-sm"
            >
              <span
                class="border rounded-full text-secondary 
      text-xs px-2 py-0.5 bg-white"
              >
                제주 맛집
              </span>
            </li>
            <li
              class="py-1 hover:bg-gray-200 px-3 text-sm"
            >
              <span
                class="border rounded-full text-secondary 
      text-xs px-2 py-0.5 bg-white"
              >
                맛집
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col"
    >
      <form
        class="flex"
      >
        <input
          class="focus:outline-none border-b px-2 py-1 text-sm md:w-80 sm:w-72 w-64 rounded-none"
          id="keywordInput"
          placeholder="#맛집, #카페, #숙소 등의 키워드를 입력해주세요 :)"
          type="text"
          value="맛집"
        />
        <button
          class="border-b text-sm text-primary w-8 h-8"
          type="submit"
        >
          <svg
            class="w-full h-full p-2 false"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8Z"
            />
            <path
              d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8Z"
            />
          </svg>
        </button>
      </form>
      <div
        class="md:w-80 sm:w-72 w-64 relative"
      >
        <ul
          class="absolute left-0 w-full border rounded-md bg-white z-50 
      top-0"
        >
          <li
            class="py-1 hover:bg-gray-200 px-3 text-sm"
          >
            <span
              class="border rounded-full text-secondary 
      text-xs px-2 py-0.5 bg-white"
            >
              제주 맛집
            </span>
          </li>
          <li
            class="py-1 hover:bg-gray-200 px-3 text-sm"
          >
            <span
              class="border rounded-full text-secondary 
      text-xs px-2 py-0.5 bg-white"
            >
              맛집
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
